### ISSUE TITLE
useRoute() does not update with current route after upgrading from 3.11.2 to 3.12.0

### ISSUE THREAD

**Issue by mkierdev** [nuxt/nuxt#28804]
### Environment

Nuxt 3.12.0
Node Version: v20.12.2

### Reproduction

```js

// pages/testPage.vue
const { testFunctionWithUseRouteInside } = useTestStore()
const route = useRoute();
console.log(route.params); // { test: 123 }
await Promise.all([testFunctionWithUseRouteInside(), otherFunc()])

// stores/testStore.ts
const testFunctionWithUseRouteInside = () => {
  const something = '' // ...
  const route = useRoute();
  console.log(route.params); // { }
  return something
}
```

### Describe the bug

Issue exists only after upgrading from 3.11.2 to 3.12 and so on.

It appears that page component has correct data inside useRoute().
However useRoute() inside store function holds 'previous route'.

Suggested workaround (https://github.com/nuxt/nuxt/issues/27199#issuecomment-2111055177) of using ``const route = useRouter().currentRoute.value;`` is not working correctly.

**Comment by fishDev94** [comment]
what's happening if you run testFunctionWithUseRouteInside() on "onMounted()" lifecycle?

**Comment by mkierdev** [comment]
> what's happening if you run testFunctionWithUseRouteInside() on "onMounted()" lifecycle?

Running in onMounted() seems fine, however above code is highly simplified :)

These functions must be SSR-friendly.
SSR is fine, navigating via client seems to be broken

**Comment by fishDev94** [comment]
i think it is a normal behaviour because if you run the code outside "onMounted" lifecycle, you are almost in "creating", it means that you are still in the old page and the new one is still mounting in progress,
it means that you are not landed yet in the new route, sometimes you can see this behaviour by using "destroy" on nuxt2 for example, it will runs after you are landed in the new route

**Comment by mkierdev** [comment]
> i think it is a normal behaviour because if you run the code outside "onMounted" lifecycle, you are almost in "creating", it means that you are still in the old page and the new one is still mounting in progress, it means that you are not landed yet in the new route, sometimes you can see this behaviour by using "destroy" on nuxt2 for example, it will runs after you are landed in the new route

Everything was working fine prior to v_3.12
I wouldn't say it's normal, code inside page component is running so the route is already changed. Function inside store is called later than useRoute inside page, which has correct route info..

**Comment by fishDev94** [comment]
but the previous route is not destroyed yet at this point i would say

**Comment by danielroe** [comment]
When navigating on client-side, there are two 'forks' of a suspense tree. One fork remains on the page displaying the old route. Inside this page the route remains the 'old' one. Another fork starts rendering the new route.

Because two routes 'exist' at the same time, we inject a `route` object inside each page that gets picked up by `useRoute` so that it 'matches' what's being rendered. Without doing this, you'd get flashing content + double mounts because the 'new' page being rendered would have a `route` that really belonged to the _old_ page, OR, you'd have new route content bleeding through into the old page that was being rendered.

However, when `useRoute` is called outside a page context, we can't know which page it's being called from, so it gets a route object that only updates after the visible page has transitioned and suspense has resolved. This is so you don't get bits of your layout shifting before the actual main page has updated.

Possibly related issues:
- https://github.com/nuxt/nuxt/issues/14595
- https://github.com/nuxt/nuxt/issues/24642

**Comment by danielroe** [comment]
Let's track in https://github.com/nuxt/nuxt/issues/28804 - see https://github.com/nuxt/nuxt/issues/28804#issuecomment-2327483434 for an explanation of what is happening and why.

**Comment by mkierdev** [comment]
> However, when `useRoute` is called outside a page context(...)

@danielroe is there any workaround available right now? Cause it prevents us from upgrading from 3.11.
Passing route from page component is DX downgrade, as functions inside store calls other functions thus route should be passed multiple times. 
Prior to 3.12 it was working fine.

**Comment by daniluk4000** [comment]
> > However, when `useRoute` is called outside a page context(...)
> 
> @danielroe is there any workaround available right now? Cause it prevents us from upgrading from 3.11. Passing route from page component is DX downgrade, as functions inside store calls other functions thus route should be passed multiple times. Prior to 3.12 it was working fine.

You can try to use my usePageRoute computed.

https://github.com/nuxt/nuxt/issues/24942#issuecomment-1895586730

Works like a charm for us

**Comment by daniluk4000** [comment]
That is true, yeah. We use this method for edge cases, but it's not a replacement for regular route sadly.

**Comment by mkierdev** [comment]
> I think the current behaviour on the Nuxt end is correct.

Yet prior to Nuxt 3.12 it was working differently and this issue is marked as bug.
Just trying to decide what to do now - is this going to be fixed or should we adapt our app to new behavior?

**Comment by avi-agg** [comment]
I am having the same issue in Nuxt 3.11.1. For me it was working fine till 3.11.0.



> Just trying to decide what to do now - is this going to be fixed or should we adapt our app to new behavior?

I have the same question.

**Comment by avi-agg** [comment]
@mkierdev please let me know if you find a solution...

**Comment by MauroGonzalez51** [comment]
Encountered the same issue working with nuxt 3.13.1, in some cases the content were updated, some others not.
So I had to use the window.location.replace for client side when the NuxtLink wasn't working.
I hope this is just a 'temporary fix'

**Comment by mkierdev** [comment]
@danielroe Are there any updates on this? Should we use useRoute() only in pages components (what for non-pages components, utilites etc)?
Currently this issue locks us on nuxt 3.11 and vue 3.4 and we are still wondering how should we approach this.

**Comment by mkierdev** [comment]
Possible workaround:
We've refactored our code to pass useRoute() from page components, so now our app is working fine with <Suspense> components.

**Comment by manafatech** [comment]
We are also encountering an issue where, after a hard page refresh, the query parameters are not being updated when the user sets them.
@danielroe when will fix this issue?

**Comment by danielroe** [comment]
@manafatech  that doesn't sound like the same issue, which is a bug related to pinia

**Comment by AcelisWeaven** [comment]
Hello,

If that helps, I've made a minimal StackBlitz reproduction of the original issue: [you can find it here](https://stackblitz.com/edit/github-tcxs4k?file=pages%2Fwith-nuxt-component%2F%5Bslug%5D.vue). The issue is triggered when navigating to the page using `defineNuxtComponent`.

Basically in our project the issue only happened in pages using `defineNuxtComponent`but we relied it to be able to run async setup functions where we needed the Nuxt context.

So here's what we did:
- Made sure all our component were using `defineNuxtComponent` (for consistency)
- We made a composable `useCurrentRoute` that replaces `useRoute`, so we can remove it easily later once this is fixed

```ts
// composables/useCurrentRoute.ts

export default function () {
  return useRouter().currentRoute.value
}
```

**Comment by BBoehm** [comment]
Another possible workaround: 

Calling useRoute() not in the beginning of the store but inside the functions once they are called solved the problem for us

**Comment by michael-scheurer** [comment]
Tested **workaround** mentioned [here](https://github.com/nuxt/nuxt/issues/21340#issuecomment-1858900454)

`
import { useRoute as useNativeRoute } from 'vue-router'
const nativeRoute = useNativeRoute()
`

Working for me with Nuxt 3.7.3

**Comment by tux2nicolae** [comment]
We have the same problem, initially we tried @BBoehm solution which didn't seem to work (we run it in `runOnce() function`, but the @michael-scheurer workaround worked for use as well.

**Comment by richgcook** [comment]
It is crazy that this used to work prior to 3.11.2. I use useRoute() is a lot of components like `<Header />` or `<Nav />` which sits outside of `<NuxtPage />` in a layout file.

What changed?

**Comment by danielroe** [comment]
this should have been resolved in https://github.com/nuxt/nuxt/pull/24673.

**Comment by daniluk4000** [comment]
> this should have been resolved in [#24673](https://github.com/nuxt/nuxt/pull/24673).

Impressive, TYSM!
